### ISSUE TITLE
AsyncMem Implementations Can Deadlock

### ISSUE THREAD

**Issue by dz333** [apl-cornell/PDL#12]
The code generator requires that memory requests can be issued if the memory module was "full" but a response is being dequeued in the same rule.

However, the current implementations don't support this and thus deadlock (the internal queue for requests is FULL and thus we can't make a request and it blocks the rule that would dequeue that request to empty the queue).

Ideally, we would modify the library so that the `req` and `resp` methods would actually execute in this scenario; however we may need a workaround.
Possible workarounds are:
 - Implement a special method that is called whenever a stage *would* do a `req` and `resp` in the same rule.
 - Modify the code generator to automatically dequeue responses a stage looking for and store them in other registers. This would break apart argument fetch and execution into two BSV rules (have to be careful to make it possible for those to execute on the same cycle to keep IPC up).

NOTE: this problem only occurs with the `AsyncMem` interface so if we only use register files, we don't have this issue (there is no queueing of requests to the register files).

---------------

The more general problem here is that, while requests to the same location happen in order, requests to different locations may be out of order. The old implementation required that responses were pulled off in the same order as requests, which lead to deadlocks.

I have reimplemented the module so that responses can be acknowledged out-of-order (i.e. any ready+valid response can be read) via a circular buffer. We still have the problem that buffers must be big enough to store all potential in-flight requests, unless we control scheduling a bit more strictly (which might be a better solution)

-----------------

Example Code that is problematic:

```
x <- mem[a];
---
y <- mem[x];
```

Basically, if we only have 1 memory port, then we can't guarantee `mem[x]` has priority over `mem[a]`
and thus queues can get filled up and we can deadlock.

One solution might be to require that such programs require _multiple read ports_.
Then we can avoid conflicts by assigning operations to different ports.

E.g.,
```
x <- mem.port1.read(a);
---
y <- mem.port2.read(x);
```

The main question becomes: what are the requirements for needing more ports? Should the original be acceptable in _any way_ or do we require that _every access_ maps uniquely to a port?

In the latter case, we'd need one port for every access in every path.

**Comment by dz333** [comment]
@eszlac Can you comment here what the current state of the Port Checker is?

I.e. what restrictions does PDL currently impose on port usage / availability?

**Comment by dz333** [comment]
This is now caught by the port checker when the following flag is enabled during compilation:

> -p, --portwarn           Throw errors for port conflicts between stages
